### PR TITLE
[AIRFLOW-5251] add missing typing-extensions dep for py37

### PR DESCRIPTION
**Pull request by houqp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Turns out Protocol is still experimental and not available in python 3.7's typing module. So typing-extensions is still required for py3.7.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by ashb** [comment]
Oh yes, Protocol isn't in core until 3.8, and even then it's under `typing`, not `typing_exetenions` :(

Could you update the build matrix so that one of the jobs runs on Python 3.7 please?

And do you think it is worth keeping this to `<3.8` and adding an import guard to the few places we use it:

```
try:
    from typing import Protocol
except ImportError:
    from typing_extensions import Protocol
```

possibly bundled up into a module somewhere in the airflow tree so that we can, for example, just do `from airflow.typing import Protocol` whereever we want to use it without worrying about the import guard everywhere.

**Comment by houqp** [comment]
Sure, I will make that change today :)

**Comment by houqp** [comment]
I have changed the mysql build to use python 3.7. @ashb @potiuk let me know if you prefer to create a new build matrix entry for 3.7 instead.

one surprise is some of the identity comparison behaviors have been changed in 3.7, so I had to put in python version check in unit test to change the assertion statement conditionally.

File: tests/utils/test_helpers.py [hunk 1]
```diff
@@ -213,7 +214,10 @@ def test_is_in(self):
         # python caches small integers, so i is 3 will be True,
         # but `big_i is 2 ** 31` is False.
         self.assertTrue(helpers.is_in(big_i, test_container))
-        self.assertFalse(helpers.is_in(2 ** 31, test_container))
+        if sys.version_info.minor < 7:
+            self.assertFalse(helpers.is_in(2 ** 31, test_container))
+        else:
+            self.assertTrue(helpers.is_in(2 ** 31, test_container))
```
Comment (ashb) [inline tests/utils/test_helpers.py]: This seems like a change in behaviour. What's going on here?

File: airflow/models/crypto.py [hunk 2]
```diff
@@ -17,8 +17,8 @@
 # specific language governing permissions and limitations
 # under the License.
 
-from typing_extensions import Protocol
-from typing import Optional
+from airflow.typing import Protocol
+from typing import Optional  # noqa
```
Comment (ashb) [inline airflow/models/crypto.py]: This shouldn't be needed as Optional is used in the type comments which counts as used for all the tools we use in CI.

```suggestion
from typing import Optional
```

File: tests/utils/test_helpers.py (hunk 1 above)
Comment (houqp) [reply tests/utils/test_helpers.py]: Yes, the big int caching behavior is changed in py 3.7. multiple creation of `2**31`s in this test is now returning the same reference id.

Comment (ashb) [reply tests/utils/test_helpers.py]: Give me a second, we don't use this helper anywhere other than tests, I'll remove it on master then this issue just goes away.

Comment (ashb) [reply tests/utils/test_helpers.py]: https://issues.apache.org/jira/browse/AIRFLOW-5276 (separate PR - you're welcome to take that one on, I wont get around to it today.)

Comment (houqp) [reply tests/utils/test_helpers.py]: I will take a stab at that today. Deleting unused code is one of my fav exercise :)

Comment (houqp) [reply tests/utils/test_helpers.py]: PR sent: https://github.com/apache/airflow/pull/5878

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
Thanks @houqp (for this and the related PR). Once this is green I'll merge.

**Comment by dstandish** [comment]
It seems this may be causing a problem with alembic.  Since you run alembic from `<root>/airflow` directory, when it tries to import `typing_extensions`, it then tries to open `typing`, which resolves to this typing module and not the actual one.

That's probably confusing, but basically if I run `alembic revision -m "something"` I get import error from tying module.

If I run from repo root (specifying config file and changing `script_location` I don't get this error.  (i get `ERROR - Multiple heads are present;`)

**Pull request merged**